test-framework/test_framework.go

```diff
@@ -0,0 +1,226 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Based on https://github.com/wk8/go-ordered-map, Copyright Jean Rougé
+ *
+ */
+
+package test
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/onflow/cadence/runtime/ast"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/errors"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"github.com/onflow/cadence/runtime/parser"
+	"github.com/onflow/cadence/runtime/sema"
+	"github.com/onflow/cadence/runtime/stdlib"
+	"github.com/onflow/cadence/runtime/tests/utils"
+)
+
+// This Provides utility methods to easily run test-scripts.
+// Example use-case:
+//   - To run all tests in a script:
+//         RunTests("source code")
+//   - To run a single test method in a script:
+//         RunTest("source code", "testMethodName")
+//
+// It is assumed that all test methods start with the 'test' prefix.
+
+const testFunctionPrefix = "test"
+
+type Results map[string]error
+
+func RunTest(script string, funcName string) error {
+	_, inter := parseCheckAndInterpret(script)
+	_, err := inter.Invoke(funcName)
+	return err
+}
+
+func RunTests(script string) Results {
+	program, inter := parseCheckAndInterpret(script)
+
+	results := make(Results)
+
+	for _, funcDecl := range program.FunctionDeclarations() {
+		funcName := funcDecl.Identifier.Identifier
+
+		if strings.HasPrefix(funcName, testFunctionPrefix) {
+			_, err := inter.Invoke(funcName)
+			results[funcName] = err
+		}
+	}
+
+	return results
+}
+
+func parseCheckAndInterpret(script string) (*ast.Program, *interpreter.Interpreter) {
+	program, err := parser.ParseProgram(script, nil)
+
+	checker, err := newChecker(program)
+	if err != nil {
+		panic(err)
+	}
+
+	err = checker.Check()
+	if err != nil {
+		panic(err)
+	}
+
+	// TODO: validate test function signature
+	//   e.g: no return values, no arguments, etc.
+
+	inter, err := newInterpreterFromChecker(checker)
+	if err != nil {
+		panic(err)
+	}
+
+	err = inter.Interpret()
+	if err != nil {
+		panic(err)
+	}
+
+	return program, inter
+}
+
+func newInterpreterFromChecker(checker *sema.Checker) (*interpreter.Interpreter, error) {
+	predeclaredInterpreterValues := stdlib.BuiltinFunctions.ToInterpreterValueDeclarations()
+	predeclaredInterpreterValues = append(predeclaredInterpreterValues, stdlib.BuiltinValues.ToInterpreterValueDeclarations()...)
+	predeclaredInterpreterValues = append(predeclaredInterpreterValues, stdlib.HelperFunctions.ToInterpreterValueDeclarations()...)
```

Comment:
Here's a utility that's helped for my encoding tests. It's overkill for a one-off like this but if this pattern is common outside of the encoding tests then it may be worth using:

```
func Concat[T any](deep ...[]T) []T {
	length := 0
	for _, b := range deep {
		length += len(b)
	}

	flat := make([]T, 0, length)
	for _, b := range deep {
		flat = append(flat, b...)
	}

	return flat
}
```

I went with iterating twice but allocating memory once. This probably isn't optimal because I didn't measure the tradeoff.

Reply:
This kind of appends only happens once during initialization usually (here it's only for tests, which are local), so allocations aren't really a problem. So using append twice would be fine/simple?